Readme.md

```diff
@@ -0,0 +1,1432 @@
+
+`Option`s returned by functions force you to check for the absence of value before using it. This can be done using pattern matching:
+
+```scala
+val v: Vector[Int] = ...
+v.find(_ > 0) match {
+  case Some(n) => println("$n is positive")
+  case None => println("No positive values")
+}
+```
+
+Moreover, `Option` has several methods which make working with possibly absent values easier:
+
+```scala
+// when we only need to handle the "value is present" case
+v.find(_ > 0).foreach { n =>
+  println("$n is positive")
+}
+
+// when we need to return a default value in case
+// the computed value is absent
+val n = v.find(_ > 0).getOrElse(0)
+```
+
+When interfacing with Java code, all values coming from the Java side must be checked for `null`s if it is not stated explicitly that no `null` values are possible. `Option.apply` methods accepts a nullable value and returns an `Option` which will be `None` if the passed value is actually `null`:
+
+```scala
+Option(javaClass.getResourceAsStream("some_file"))
+```
+
+Do not call `get` on optional values except when you're absolutely sure that the optional value is always non-empty. Often such code may be rewritten more clearly using pattern matching or other optional combinators.
+
+### Recursion
+
+Recursion is often very natural for certain algorithms, especially on recursive structures like trees and lists. In many cases, however, recursion tends to make the code obscure.
+
+Always try to solve the problem without explicit recursion first. Scala collections library provides lots of combinators which make explicit recursive traversals unnecessary in many cases.
+
+If a problem cannot be solved with pre-defined, combinators, compare the imperative version using loops and recursion first. In many cases, especially when the logic is not very large, imperative implementation may be simpler than recursion. In other cases, recursion would be clearer.
+
+When writing recursive functions, try to write them in a tail-recursive way, enforced by the `@tailrec` annotation. Tail-recursive functions will be optimized by the compiler into an imperative loop. However, tail-recursive functions may require passing lots of state in an accumulator parameter. Sometimes rewriting such functions using imperative loops may make the code nicer.
+
+## Object-oriented programming
+
+Scala is a powerful object-oriented language; if fact, it is even more pure than Java. Everything in Scala is an object, and its facilities for object-oriented design are very expressive and convenient. Naturally, most of object-oriented design practices are possible in Scala; they are described in numerous forms in various books and articles and therefore are omitted here. In this document we will examine Scala-specific features which help with the object-oriented programming.
+
+### Dependency injection
```

Comment:
How about the following view on DI?

DI should be viewed as a way to facilitate unit testing and explicitly list dependencies, not as a way to construct the entire dependency graph for the entire process. This means that rather than having a single assembly point, each module should construct its own subgraph. "A module" is whatever you may want to reuse a test in isolation at a particular abstraction level: "manifest validator", "database connection", "web portal" are all modules. 

One disadvantage of this approach is that you won't be able to replace, say, a shared data access tier once and for everyone.